File: leetcode/0545-boundary-of-binary-tree/solution.cpp
#include <vector>
#include <algorithm>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> boundaryOfBinaryTree(TreeNode* root) {
        vector<int> result;
        if (root == nullptr) {
            return result;
        }

        if (!isLeaf(root)) {
            result.push_back(root->val);
        }

        TreeNode* node = root->left;
        while (node != nullptr) {
            if (!isLeaf(node)) {
                result.push_back(node->val);
            }
            node = node->left != nullptr ? node->left : node->right;
        }

        collectLeaves(root, result);

        vector<int> rightBoundary;
        node = root->right;
        while (node != nullptr) {
            if (!isLeaf(node)) {
                rightBoundary.push_back(node->val);
            }
            node = node->right != nullptr ? node->right : node->left;
        }
        reverse(rightBoundary.begin(), rightBoundary.end());
        result.insert(result.end(), rightBoundary.begin(), rightBoundary.end());

        return result;
    }

private:
    bool isLeaf(TreeNode* node) {
        return node->left == nullptr && node->right == nullptr;
    }

    void collectLeaves(TreeNode* node, vector<int>& result) {
        if (node == nullptr) {
            return;
        }
        if (isLeaf(node)) {
            result.push_back(node->val);
            return;
        }
        collectLeaves(node->left, result);
        collectLeaves(node->right, result);
    }
};

File: leetcode/0545-boundary-of-binary-tree/Solution.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> boundaryOfBinaryTree(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        if (!isLeaf(root)) {
            result.add(root.val);
        }

        TreeNode node = root.left;
        while (node != null) {
            if (!isLeaf(node)) {
                result.add(node.val);
            }
            node = node.left != null ? node.left : node.right;
        }

        collectLeaves(root, result);

        List<Integer> rightBoundary = new ArrayList<>();
        node = root.right;
        while (node != null) {
            if (!isLeaf(node)) {
                rightBoundary.add(node.val);
            }
            node = node.right != null ? node.right : node.left;
        }
        Collections.reverse(rightBoundary);
        result.addAll(rightBoundary);

        return result;
    }

    private boolean isLeaf(TreeNode node) {
        return node.left == null && node.right == null;
    }

    private void collectLeaves(TreeNode node, List<Integer> result) {
        if (node == null) {
            return;
        }
        if (isLeaf(node)) {
            result.add(node.val);
            return;
        }
        collectLeaves(node.left, result);
        collectLeaves(node.right, result);
    }
}
